prow/plugins/bugzilla/bugzilla.go

```diff
@@ -0,0 +1,389 @@
+	GetPullRequest(org, repo string, number int) (*github.PullRequest, error)
+	CreateComment(owner, repo string, number int, comment string) error
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	AddLabel(owner, repo string, number int, label string) error
+	RemoveLabel(owner, repo string, number int, label string) error
+}
+
+func handleGenericComment(pc plugins.Agent, e github.GenericCommentEvent) error {
+	event, err := digestComment(pc.GitHubClient, pc.Logger, e)
+	if err != nil {
+		return err
+	}
+	if event != nil {
+		options := pc.PluginConfig.Bugzilla.OptionsForBranch(event.org, event.repo, event.baseRef)
+		return handle(*event, pc.GitHubClient, pc.BugzillaClient, options, pc.Logger)
+	}
+	return nil
+}
+
+func handlePullRequest(pc plugins.Agent, pre github.PullRequestEvent) error {
+	event, err := digestPR(pc.Logger, pre)
+	if err != nil {
+		return err
+	}
+	if event != nil {
+		options := pc.PluginConfig.Bugzilla.OptionsForBranch(event.org, event.repo, event.baseRef)
+		return handle(*event, pc.GitHubClient, pc.BugzillaClient, options, pc.Logger)
+	}
+	return nil
+}
+
+// digestPR determines if any action is necessary and creates the objects for handle() if it is
+func digestPR(log *logrus.Entry, pre github.PullRequestEvent) (*event, error) {
+	// These are the only actions indicating the PR title may have changed.
+	if pre.Action != github.PullRequestActionOpened &&
+		pre.Action != github.PullRequestActionReopened &&
+		pre.Action != github.PullRequestActionEdited {
+		return nil, nil
+	}
+
+	var (
+		org     = pre.PullRequest.Base.Repo.Owner.Login
+		repo    = pre.PullRequest.Base.Repo.Name
+		baseRef = pre.PullRequest.Base.Ref
+		number  = pre.PullRequest.Number
+		title   = pre.PullRequest.Title
+	)
+
+	// Make sure the PR title is referencing a bug
+	e := &event{org: org, repo: repo, baseRef: baseRef, number: number, body: title, htmlUrl: pre.PullRequest.HTMLURL, login: pre.PullRequest.User.Login}
+	mat := titleMatch.FindStringSubmatch(title)
+	if mat == nil {
+		// in the case that the title used to reference a bug and no longer does we
+		// want to handle this to remove labels
+		e.missing = true
+	} else {
+		id, err := strconv.Atoi(mat[1])
+		if err != nil {
+			// should be impossible based on the regex
+			log.WithError(err).Debug("Failed to parse bug ID as int - is the regex correct?")
+			return nil, err
+		}
+		e.bugId = id
+	}
+
+	// Check if the previous version of the title referenced a bug.
+	var changes struct {
+		Title struct {
+			From string `json:"from"`
+		} `json:"title"`
+	}
+	if err := json.Unmarshal(pre.Changes, &changes); err != nil {
+		// we're detecting this best-effort so we can handle it anyway
+		if e.missing {
+			// we only want to expose missing if the previous title had a bug
+			return nil, nil
+		}
+		return e, nil
+	}
+	prevMat := titleMatch.FindStringSubmatch(changes.Title.From)
+	if prevMat == nil {
+		// title did not previously reference a bug
+		return nil, nil
```

Comment:
If the title did not previously reference a bug, but does now we need to handle the event.

Reply:
eek, nice catch